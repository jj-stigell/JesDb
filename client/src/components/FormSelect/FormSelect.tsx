import React, { ReactNode } from 'react'
import { Select } from 'flowbite-react'

interface IFormSelect {
  id: string
  disabled: boolean
  options: string[]
  labelText: string | ReactNode
  helperText: string | ReactNode
}

export default function FormSelect ({
  id,
  disabled,
  options,
  labelText,
  helperText
}: IFormSelect): React.JSX.Element {
  return (
    <div>
      <label
        htmlFor={id}
        className="block mb-2 text-sm text-gray-900 dark:text-white"
      >
        {labelText}
      </label>
      <Select
        id={id}
        disabled={disabled}
        aria-describedby={`helper-text-${id}`}
      >
        <option value="" disabled>
          Select an option
        </option>
        {options.map((fieldName: string) => (
          <option key={`${id}_${fieldName}`} value={fieldName}>
            {fieldName}
          </option>
        ))}
      </Select>
      <p
        id={`helper-text-${id}`}
        className="mt-2 text-sm text-gray-500 dark:text-gray-200"
      >
        {helperText}
      </p>
    </div>
  )
}
