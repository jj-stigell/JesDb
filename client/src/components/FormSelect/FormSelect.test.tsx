import React from 'react'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import FormSelect from './FormSelect'

describe('FormSelect Component', () => {
  const defaultProps = {
    id: 'test-select',
    disabled: false,
    options: ['Option 1', 'Option 2', 'Option 3'],
    labelText: 'Select an option',
    helperText: 'Helper text'
  }

  it('should render the component with all props', () => {
    const { asFragment } = render(<FormSelect {...defaultProps} />)

    expect(asFragment()).toMatchSnapshot()
    expect(screen.getByLabelText('Select an option')).toBeInTheDocument()
    expect(screen.getByText('Helper text')).toBeInTheDocument()
    defaultProps.options.forEach(option => {
      expect(screen.getByText(option)).toBeInTheDocument()
    })
  })

  it('should display the label text correctly', () => {
    render(<FormSelect {...defaultProps} />)

    expect(screen.getByLabelText('Select an option')).toBeInTheDocument()
  })

  it('should display the helper text correctly', () => {
    render(<FormSelect {...defaultProps} />)

    expect(screen.getByText('Helper text')).toBeInTheDocument()
  })

  it('should render the correct number of options', () => {
    render(<FormSelect {...defaultProps} />)
    const options = screen.getAllByRole('option')

    // +1 for the default "-" option (omit)
    expect(options).toHaveLength(defaultProps.options.length + 1)
  })

  it('should disable the select element when disabled prop is true', () => {
    const { asFragment } = render(<FormSelect {...defaultProps} disabled={true} />)

    expect(asFragment()).toMatchSnapshot()
    expect(screen.getByLabelText('Select an option')).toBeDisabled()
  })

  it('should enable the select element when disabled prop is false', () => {
    render(<FormSelect {...defaultProps} disabled={false} />)

    expect(screen.getByLabelText('Select an option')).toBeEnabled()
  })
})
