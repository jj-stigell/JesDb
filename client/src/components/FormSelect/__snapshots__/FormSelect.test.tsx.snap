// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSelect Component should disable the select element when disabled prop is true 1`] = `
<DocumentFragment>
  <div>
    <label
      class="block mb-2 text-sm text-gray-900 dark:text-white"
      for="test-select"
    >
      Select an option
    </label>
    <div
      class="flex"
    >
      <div
        class="relative w-full"
      >
        <select
          aria-describedby="helper-text-test-select"
          class="block w-full border disabled:cursor-not-allowed disabled:opacity-50 border-gray-300 bg-gray-50 text-gray-900 focus:border-cyan-500 focus:ring-cyan-500 dark:border-gray-600 dark:bg-gray-700 dark:text-white dark:placeholder-gray-400 dark:focus:border-cyan-500 dark:focus:ring-cyan-500 p-2.5 text-sm rounded-lg"
          disabled=""
          id="test-select"
        >
          <option
            disabled=""
            value=""
          >
            Select an option
          </option>
          <option
            value="Option 1"
          >
            Option 1
          </option>
          <option
            value="Option 2"
          >
            Option 2
          </option>
          <option
            value="Option 3"
          >
            Option 3
          </option>
        </select>
      </div>
    </div>
    <p
      class="mt-2 text-sm text-gray-500 dark:text-gray-200"
      id="helper-text-test-select"
    >
      Helper text
    </p>
  </div>
</DocumentFragment>
`;

exports[`FormSelect Component should render the component with all props 1`] = `
<DocumentFragment>
  <div>
    <label
      class="block mb-2 text-sm text-gray-900 dark:text-white"
      for="test-select"
    >
      Select an option
    </label>
    <div
      class="flex"
    >
      <div
        class="relative w-full"
      >
        <select
          aria-describedby="helper-text-test-select"
          class="block w-full border disabled:cursor-not-allowed disabled:opacity-50 border-gray-300 bg-gray-50 text-gray-900 focus:border-cyan-500 focus:ring-cyan-500 dark:border-gray-600 dark:bg-gray-700 dark:text-white dark:placeholder-gray-400 dark:focus:border-cyan-500 dark:focus:ring-cyan-500 p-2.5 text-sm rounded-lg"
          id="test-select"
        >
          <option
            disabled=""
            value=""
          >
            Select an option
          </option>
          <option
            value="Option 1"
          >
            Option 1
          </option>
          <option
            value="Option 2"
          >
            Option 2
          </option>
          <option
            value="Option 3"
          >
            Option 3
          </option>
        </select>
      </div>
    </div>
    <p
      class="mt-2 text-sm text-gray-500 dark:text-gray-200"
      id="helper-text-test-select"
    >
      Helper text
    </p>
  </div>
</DocumentFragment>
`;
